File: gims_mobile_app/lib/models/fire/fire_policy.dart
import '../utility_models.dart';

class FirePolicy {
  int? id;
  String? sysNumber;
  InsuranceCompany? company;
  DateTime? date;
  Bank? bank;
  Branch? branch;
  String? policyholder;
  String? address;
  String? stockInsured;
  double? sumInsured;
  String? interestInsured;
  String? coverage;
  String? location;
  String? construction;
  String? owner;
  String? usedAs;
  DateTime? periodFrom;
  DateTime? periodTo;
  bool isHovered = false;

  FirePolicy({
    this.id,
    this.sysNumber,
    this.company,
    this.date,
    this.bank,
    this.branch,
    this.policyholder,
    this.address,
    this.stockInsured,
    this.sumInsured,
    this.interestInsured,
    this.coverage,
    this.location,
    this.construction,
    this.owner,
    this.usedAs,
    this.periodFrom,
    this.periodTo,
  });

  factory FirePolicy.fromJson(Map<String, dynamic> json) {
    // Helper to safely parse doubles (especially for large numbers/scientific notation)
    double? parseDouble(dynamic value) {
      if (value == null) return null;
      if (value is num) return value.toDouble();
      if (value is String) return double.tryParse(value);
      return null;
    }

    return FirePolicy(
      id: json['policy_id'] ?? json['id'],
      sysNumber: json['sys_number']?.toString(),
      company: json['company'] != null ? InsuranceCompany.fromJson(json['company']) : null,
      date: json['date'] != null ? DateTime.tryParse(json['date'].toString()) : null,
      bank: json['bank'] != null ? Bank.fromJson(json['bank']) : null,
      branch: json['branch'] != null ? Branch.fromJson(json['branch']) : null,
      policyholder: json['policyholder']?.toString() ?? 'N/A',
      address: json['address']?.toString() ?? '',
      stockInsured: json['stock_insured']?.toString() ?? '',
      sumInsured: parseDouble(json['sum_insured']),
      interestInsured: json['interest_insured']?.toString() ?? '',
      coverage: json['coverage']?.toString() ?? '',
      location: json['location']?.toString() ?? '',
      construction: json['construction']?.toString() ?? '',
      owner: json['owner']?.toString() ?? '',
      usedAs: json['used_as']?.toString() ?? '',
      periodFrom: json['period_from'] != null ? DateTime.tryParse(json['period_from'].toString()) : null,
      periodTo: json['period_to'] != null ? DateTime.tryParse(json['period_to'].toString()) : null,
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (id != null) data['id'] = id;
    if (sysNumber != null) data['sysNumber'] = sysNumber;
    if (company != null) data['company'] = {'id': company!.id};
    if (date != null) data['date'] = date?.toIso8601String();
    if (bank != null) data['bank'] = {'id': bank!.id};
    if (branch != null) data['branch'] = {'id': branch!.id};
    data['policyholder'] = policyholder;
    data['address'] = address;
    data['stockInsured'] = stockInsured;
    data['sumInsured'] = sumInsured;
    data['interestInsured'] = interestInsured;
    data['coverage'] = coverage;
    data['location'] = location;
    data['construction'] = construction;
    data['owner'] = owner;
    data['usedAs'] = usedAs;
    if (periodFrom != null) data['periodFrom'] = periodFrom?.toIso8601String();
    if (periodTo != null) data['periodTo'] = periodTo?.toIso8601String();
    return data;
  }
}

File: gims_backend_api/src/main/java/com/kutub/InsuranceManagement/entity/fire/FirePolicy.java
package com.kutub.InsuranceManagement.entity.fire;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.kutub.InsuranceManagement.entity.utility.BnkInfo;
import com.kutub.InsuranceManagement.entity.utility.BnkBrInfo;
import com.kutub.InsuranceManagement.entity.utility.InsInfo;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "fire_policy")
public class FirePolicy {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonProperty("policyId")
    @JsonAlias("id")
    private int id;

    @Column(name = "sys_number", unique = true, nullable = true)
    private String sysNumber;

    @ManyToOne
    @JoinColumn(name = "company_id", nullable = false)
    private InsInfo company;

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date date = new Date();

    @ManyToOne
    @JoinColumn(name = "bank_id", nullable = false)
    private BnkInfo bank;

    @ManyToOne
    @JoinColumn(name = "branch_id", nullable = true)
    private BnkBrInfo branch;

    @Column(nullable = false)
    private String policyholder;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false)
    private String stockInsured;

    @Column(nullable = false)
    private double sumInsured;

    @Column(nullable = false)
    private String interestInsured;

    @Column(nullable = false)
    private String coverage;

    @Column(nullable = false)
    private String location;

    @Column(nullable = false)
    private String construction;

    @Column(nullable = false)
    private String owner;

    @Column(nullable = false)
    private String usedAs;

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date periodFrom;

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date periodTo;

    @JsonIgnore
    @OneToMany(mappedBy = "policy", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    private List<FireBill> bills;

    // Custom setter for periodFrom
    public void setPeriodFrom(Date periodFrom) {
        this.periodFrom = periodFrom;
        setPeriodToAutomatically();
    }

    // Automatically sets periodTo to one year after periodFrom
    private void setPeriodToAutomatically() {
        if (this.periodFrom != null) {
            Calendar calendar = Calendar.getInstance();
            calendar.setTime(this.periodFrom);
            calendar.add(Calendar.YEAR, 1);
            this.periodTo = calendar.getTime();
        }
    }
}
